---
title: "R Notebook"
output: html_notebook
---

#Load packages
```{r}
library(dplyr)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(ggplot2)
library(viridisLite)
library(patchwork)
library(future)
plan('multisession', workers = 5)#cores = 5
options(future.globals.maxSize = 100000 * 1024^2)

```

#Load data
```{r}
# Load the 6mo dataset
mon6.data <- Read10X_h5("/home/pheonix/Documents/h5files/6monthpig.h5")

# Initialize the Seurat object with the raw (non-normalized data).
Mon6 <- CreateSeuratObject(counts = mon6.data, project = "6mo", min.cells = 3, min.features = 200)

Mon6
```

#QC
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
Mon6[["percent.mt"]] <- PercentageFeatureSet(Mon6, pattern = "^MT")

# Visualize QC metrics as a violin plot
VlnPlot(Mon6, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(Mon6, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Mon6, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

#QC filtering
```{r}
Mon6#before QC filter

VlnPlot(Mon6, features = 'nCount_RNA') + coord_cartesian(ylim = c(0, 10000))
VlnPlot(Mon6, features = 'nFeature_RNA') + coord_cartesian(ylim = c(0, 10000))

Mon6 <- subset(Mon6, subset = 
                 nFeature_RNA > 1000 &
                 nFeature_RNA < 10000 & 
                 percent.mt < 5)

Mon6
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
Mon6 <- SCTransform(Mon6, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Mon6 <- RunPCA(Mon6, verbose = FALSE)
ElbowPlot(Mon6, ndims = 50)
toc()
```

## UMAP (initial)
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
Mon6 <- RunUMAP(Mon6, dims = 1:40, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Mon6 <- FindNeighbors(Mon6, dims = 1:40, verbose = FALSE)
Mon6 <- FindClusters(Mon6, verbose = FALSE, algorithm = 3)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Mon6, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6, label = FALSE, pt.size = 1.0) + NoLegend()
#DimPlot(Mon6, label = FALSE)#no numbers on clusters, legend off to side

toc()
```

## FeaturePlots: view single-cell gene expression across clusters
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
GOI <- c('nCount_RNA',
         'PDGFRA', 'VIM', 'ACTA2', 'LEF1', 'CRABP2',
         'CRABP1', 'RGS5', 'PECAM1', 'LYVE1', 'FLT4', 
         'PTPRC', 'CD86', 'NKG7', 'CDH1',
         'ITGB4', 'ITGA6', 'SOX6', 'KRT15', 'KRT14', 'KRT5', 'KRT1', 'KRT10', 'KRT8', 'CDSN', 
         'TOP2A', 'SOX10', 'DCT', 'SOX9', 'DLK2', 'DLL1', 'PDGFA', 'VEGFA', 'ANGPTL1', 'ANGPTL2')
#path_out <- '.'
## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Mon6,
                    features = GOI[n_plot],
                    pt.size = 1,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
  }
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
Mon6 <- FindClusters(Mon6, verbose = FALSE, algorithm = 3, resolution = 0.7)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Mon6, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

## Differential Gene Expression
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/6mo/'#path to export folder
COI <- levels(Mon6)#define the clusters you wish to know diff. gene expression of

n_genes <- length(Mon6@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(Mon6, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, (n_loops-1), '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

## Label the Clusters by Cell Type
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids <- c('0 Pericyte', '1 Basal Krtno', '2 Diff Krtno', '3 Fibroblast', '4 Diff Krtno',
                     '5 Diff Krtno', '6 BV', '7 Diff Krtno', '8 Div Krtno', '9 Immune Cell',
                     '10 Macrophage', '11 CDH1 Krtno', '12 SG', '13 Immune Cell', '14 Macrophage',
                     '15 Schwann Cell', '16 LV', '17 Corneocyte')
names(new.cluster.ids) <- levels(Mon6)
Mon6 <- RenameIdents(Mon6, new.cluster.ids)
```

Save the Seurat object
```{r}
save(Mon6, file = '/media/sean/T2TB/Object/6mo_pig_allclusters_v2.RData')
```


## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(Mon6, label = FALSE, pt.size = 0.5, 
              cols = c('#7fcdbb', '#002783', '#64a7e9', '#238b45', '#4d8ad5',
                     '#96ccff', '#8c6bb1', '#0c3695', '#2565b5', '#BB0000',
                     '#790001', '#16539e', '#00206a', '#dc0000', '#900101',
                     '#ffab4b', '#88419d', '#9decff')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Pig 6mo UMAP (All Clusters).png', path = path_out,
         width = 7, height = 7, units = 'in')
```

# 2. Subset the Keratinocytes
```{r message=FALSE, warning=FALSE}
Mon6_krtno <- subset(Mon6, idents = c('1 Basal Krtno', '2 Krtno', '4 Krtno', '5 K1_K10 Krtno', '7 Krtno', '8 Div Krtno', 
                     '11 CDH1 Krtno', '12 Basal Krtno', '17 Corneocyte'), invert = FALSE)#subset the keratinocyte supercluster
Mon6_krtno#view the subset

## We want to renormalize the data to bring out the heterogenetiy within the lineage
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
Mon6_krtno <- SCTransform(Mon6_krtno, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Mon6_krtno <- RunPCA(Mon6_krtno, verbose = FALSE)
ElbowPlot(Mon6_krtno, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
Mon6_krtno <- RunUMAP(Mon6_krtno, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Mon6_krtno <- FindNeighbors(Mon6_krtno, dims = 1:30, verbose = FALSE)
Mon6_krtno <- FindClusters(Mon6_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(Mon6_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
Includes some markers identified by Sole-Boldo et al. in their analysis as well as DAGs from Thompson et al. 2022
```{r fig.height=5, fig.width=7}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(Mon6_krtno, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('PDGFRA', 'APOE', 'KRT14', 'KRT5', 'KRT10', 'KRT18', 'KRT8', 'CALML5', 'CDSN', 'CNFN', 'CDH1',
         'TOP2A', 'ASPM', 'ITGA6', 'ITGB1', 'SOX6', 'SOSTDC1', 'KRT15', 'COL17A1', 'SOX9', 'VEGFA', 'KRT10', 'nFeature_SCT', 'TGFBI', 'DPP4', 'ACTA2', 'PPARG', 'LEF1',
         'DLL1', 'CAV1', 'DUSP6', 'COL17A1', 'CTNNB1')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Mon6_krtno,
                    features = GOI[n_plot],
                    pt.size = 1.0,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
Mon6_krtno <- FindClusters(Mon6_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Mon6_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

# Differential Gene Expression Within Keratinocytes
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/Krtno/6mo/'#path to export folder
COI <- c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10', '11')#define the clusters you wish to know diff. gene expression of

n_genes <- length(Mon6_krtno@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(Mon6_krtno, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, COI[n_loops], '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

## Label the clusters
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.sub <- c('0 Diff Krtno', '1 Diff Krtno', '2 Basal Krtno', '3 Diff Krtno', 
                         '4 Basal Krtno', '5 Basal Krtno', '6 Div Krtno', '7 AREG Krtno', 
                         '8 Sweat Gland', '9 SOX9 Krtno', '10 Corneocyte', '11 Basal Krtno')
names(new.cluster.ids.sub) <- levels(Mon6_krtno)
Mon6_krtno <- RenameIdents(Mon6_krtno, new.cluster.ids.sub)
```

##Fig Quality DimPlot
```{r fig.height=5, fig.width=7}
DimPlot(Mon6_krtno, label = TRUE, pt.size = 1.0, label.size = 6,
        cols = c('#f09527', '#f0b627', '#107de9', '#ff692b', 
                 '#2c72c8', '#0c3695', '#225ea8', '#8c6bb1', 
                 '#1a9850', '#00206a', '#d04800', '#6b92c3'))#legend
DimPlot(Mon6_krtno, label = FALSE, pt.size = 1.0,
        cols = c('#f09527', '#f0b627', '#107de9', '#ff692b', 
                 '#2c72c8', '#0c3695', '#225ea8', '#8c6bb1', 
                 '#1a9850', '#00206a', '#d04800', '#6b92c3')) + NoLegend()#blank

```

# 2b. Subset the Keratinocytes Only
```{r message=FALSE, warning=FALSE}
Mon6_krtno_sub <- subset(Mon6_krtno, idents = c('0 Diff Krtno', '1 Diff Krtno', '2 Basal Krtno', '3 Diff Krtno', 
                         '4 Basal Krtno', '5 Basal Krtno', '6 Div Krtno', '7 AREG Krtno', '9 SOX9 Krtno', 
                         '10 Corneocyte', '11 Basal Krtno'), invert = FALSE)#subset the keratinocyte supercluster
Mon6_krtno_sub#view the subset

## We want to renormalize the data to bring out the heterogenetiy within the lineage
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
Mon6_krtno_sub <- SCTransform(Mon6_krtno_sub, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
Mon6_krtno_sub <- RunPCA(Mon6_krtno_sub, verbose = FALSE)
ElbowPlot(Mon6_krtno_sub, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
Mon6_krtno_sub <- RunUMAP(Mon6_krtno_sub, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
Mon6_krtno_sub <- FindNeighbors(Mon6_krtno_sub, dims = 1:30, verbose = FALSE)
Mon6_krtno_sub <- FindClusters(Mon6_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(Mon6_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
ENSSSCG00000026302 (MKI67)
```{r fig.height=5, fig.width=7}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(Mon6_krtno_sub, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('KRT15', 'KRT14', 'KRT5', 'KRT10', 'KRT1', 'KRT8', 'KRT18', 'CALML5', 'CNFN', 'CDH1', 'LRIG1',
         'ENSSSCG00000026302', 'ITGA6', 'ITGB4', 'SOX6', 'SOX5', 'SOSTDC1', 'COL17A1', 'SOX9', 'VEGFA', 'ANGPTL4', 'PDGFC', 'TGFBI', 'TGFB2', 'CXCL14', 
         'LEF1', 'DLK2', 'DLL1', 'JAG1', 'NOTCH3',
         'DUSP6', 'CAV1', 'ASS1', 'VIM', 'PTPRC', 'PDGFRA', 'percent.mt', 'nCount_SCT', 'AREG', 'SBSN', 'AXIN2', 'WNT16')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = Mon6_krtno_sub,
                    features = GOI[n_plot],
                    pt.size = 0.75,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
Mon6_krtno_sub <- FindClusters(Mon6_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.5)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(Mon6_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(Mon6_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

# Differential Gene Expression Within Keratinocytes
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Sean/DEG/Krtno_sub/6mo_v2/'#path to export folder
COI <- c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10')#define the clusters you wish to know diff. gene expression of

n_genes <- length(Mon6_krtno_sub@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(Mon6_krtno_sub, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, COI[n_loops], '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

## Label the clusters
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.sub <- c('0 Diff Krtno', '1 Diff Krtno', '2 Suprabasal Krtno', '3 Diff Krtno', 
                         '4 Suprabasal Krtno', '5 Basal Krtno', '6 Div Krtno', '7 Diff Krtno', 
                         '8 AREG Krtno', '9 SOX9 Krtno', '10 Suprabasal Krtno')
names(new.cluster.ids.sub) <- levels(Mon6_krtno_sub)
Mon6_krtno_sub <- RenameIdents(Mon6_krtno_sub, new.cluster.ids.sub)
```

Save the Seurat object
```{r}
save(Mon6_krtno_sub, file = '/media/sean/T2TB/Object/6mo_pig_krtno_sub_v2.RData')
```

##Fig Quality DimPlot
```{r fig.height=7, fig.width=7}
DimPlot(Mon6_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6,
        cols = c('#f0b627', '#ff692b', '#2c72c8', '#d04800', '#107de9',
                 '#0c3695', '#225ea8', '#f09527', '#ffab4b', '#00206a',
                 '#2353c1'))#legend
DimPlot(Mon6_krtno_sub, label = FALSE, pt.size = 1.0,
        cols = c('#f0b627', '#ff692b', '#2c72c8', '#d04800', '#107de9',
                 '#0c3695', '#225ea8', '#f09527', '#ffab4b', '#00206a',
                 '#2353c1')) + NoLegend()#blank
```

