---
title: "R Notebook"
output: html_notebook
---

#Load packages
```{r}
library(dplyr)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(monocle3)
library(ggplot2)
library(viridisLite)
library(patchwork)
library(future)
plan('multisession', workers = 5)#cores = 5
options(future.globals.maxSize = 100000 * 1024^2)

```

#Load data
```{r}
# Load the D10 dataset
d10.data <- Read10X_h5("/home/pheonix/Documents/h5files/p10pig.h5")
#mon6.data <- Read10X_h5("/home/pheonix/Documents/h5files/6monthpig.h5")

# Initialize the Seurat object with the raw (non-normalized data).
D10 <- CreateSeuratObject(counts = d10.data, project = "P10", min.cells = 1, min.features = 500)

D10
```

#QC
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
D10[["percent.mt"]] <- PercentageFeatureSet(D10, pattern = "^MT")

# Visualize QC metrics as a violin plot
VlnPlot(D10, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(D10, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(D10, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

#QC filtering
```{r}
D10#before QC filter

D10 <- subset(D10, subset = nFeature_RNA < 6000 & percent.mt < 5)

D10
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
D10 <- SCTransform(D10, vars.to.regress = c("percent.mt", "nCount_RNA"), verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
D10 <- RunPCA(D10, verbose = FALSE)
ElbowPlot(D10, ndims = 50)
toc()
```

## UMAP (initial)
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
D10 <- RunUMAP(D10, dims = 1:40, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
D10 <- FindNeighbors(D10, dims = 1:40, verbose = FALSE)
D10 <- FindClusters(D10, verbose = FALSE, algorithm = 3)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(D10, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(D10, label = FALSE, pt.size = 1.0) + NoLegend()
#DimPlot(D10, label = FALSE)#no numbers on clusters, legend off to side

toc()
```

## FeaturePlots: view single-cell gene expression across clusters
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
GOI <- c('nCount_RNA',
         'PDGFRA', 'LRRC17', 'RUNX1T1', 'VIM', 'ACTA2', 
         'CRABP1', 'ALX4', 'LEF1', 'RGS5', 'PECAM1', 'LYVE1', 'FLT4', 'PTPRC', 'CD86', 
         'CDH1', 'KRT14', 'KRT5', 'KRT1', 'KRT10', 'KRT8', 'CDSN', 'SOX6', 'SOX9',
         'TOP2A', 'ITGA6', 'PPARG', 'FABP4', 'COCH', 'BMP7', 'VEGFA', 'SOX10', 'COL11A1')
#path_out <- '.'
## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = D10,
                    features = GOI[n_plot],
                    pt.size = 1.0,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
  }
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
D10 <- FindClusters(D10, verbose = FALSE, algorithm = 3, resolution = 0.5)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(D10, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(D10, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

## Differential Gene Expression
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/P10/'#path to export folder
COI <- levels(D10)#define the clusters you wish to know diff. gene expression of

n_genes <- length(D10@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(D10, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, (n_loops-1), '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

## Label the Clusters by Cell Type
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids <- c('0 Diff Krtno', '1 Diff Krtno', '2 Diff Krtno', '3 Basal Krtno', '4 Basal Krtno',
                     '5 RF', '6 SG', '7 Basal Krtno', '8 Div Krtno', '9 Pericyte',
                     '10 BV', '11 PF', '12 Immune Cell', '13 Sw', '14 Fibro', 
                     '15 ACTA2 Fibro', '16 Schwann Cell', '17 LV')
names(new.cluster.ids) <- levels(D10)
D10 <- RenameIdents(D10, new.cluster.ids)
```

Save the Seurat object
```{r}
save(D10, file = '/media/sean/T2TB/Object/p10_pig_allclusters_v3.RData')
```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Sean/UMAP/'
p1 <- DimPlot(D10, label = FALSE, pt.size = 1.0, 
              cols = c('#4d8ad5', '#96ccff', '#64a7e9', '#002783', '#16539e',
                     '#238b45', '#00206a', '#0c3695', '#2565b5', '#7fcdbb',
                     '#8c6bb1', '#00441b', '#900101', '#fc8d59', '#41ab5d', 
                     '#006837', '#ffab4b', '#88419d')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Pig P10 UMAP (All Clusters) updated.png', path = path_out,
         width = 7, height = 7, units = 'in')
```

# Subset the Keratinocytes
```{r message=FALSE, warning=FALSE}
P10_krtno <- subset(D10, idents = c('0 Diff Krtno', '1 Diff Krtno', '2 Diff Krtno', '3 Basal Krtno', '4 Basal Krtno',
                     '6 SG', '7 Basal Krtno', '8 Div Krtno', '13 Sw'), invert = FALSE)#subset the keratinocyte supercluster
P10_krtno#view the subset

## We want to renormalize the data to bring out the heterogenetiy within the lineage
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
P10_krtno <- SCTransform(P10_krtno, vars.to.regress = "percent.mt", verbose = FALSE)
toc()
#Dimensional Reduction and Elbow Plot
P10_krtno <- RunPCA(P10_krtno, verbose = FALSE)
ElbowPlot(P10_krtno, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
P10_krtno <- RunUMAP(P10_krtno, dims = 1:20, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
P10_krtno <- FindNeighbors(P10_krtno, dims = 1:20, verbose = FALSE)
P10_krtno <- FindClusters(P10_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(P10_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P10_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
Includes some markers identified by Sole-Boldo et al. in their analysis as well as DAGs from Thompson et al. 2022
```{r fig.height=5, fig.width=7}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(P10_krtno, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('PDGFRA', 'APOE', 'KRT14', 'KRT5', 'KRT10', 'KRT18', 'KRT8', 'CALML5', 'CDSN', 'CNFN', 'CDH1',
         'TOP2A', 'ITGA6', 'ITGB1', 'SOX6', 'SOX9', 'KRT15', 'VEGFA', 'nFeature_SCT', 'TGFBI', 'DPP4', 'ACTA2', 'PPARG', 'LEF1',
         'DLL1', 'CAV1', 'DUSP6', 'COL17A1', 'CTNNB1')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P10_krtno,
                    features = GOI[n_plot],
                    pt.size = 1.0,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
P10_krtno <- FindClusters(P10_krtno, verbose = FALSE, algorithm = 3, resolution = 0.15)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P10_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P10_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

# Differential Gene Expression Within Keratinocytes
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Desktop/Sean_CellChat/DEG/Krtno/P10/'#path to export folder
COI <- c('0', '1', '2', '3', '4', '5', '6', '7')#define the clusters you wish to know diff. gene expression of

n_genes <- length(P10_krtno@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(P10_krtno, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, COI[n_loops], '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

## Label the clusters
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.sub <- c('0 Diff Krtno', '1 Diff Krtno', '2 Diff Krtno', '3 Basal Krtno', '4 PRICKLE2 Krtno', '5 Div Krtno', '6 Sweat Gland', '7 KRT8_18 Krtno')
names(new.cluster.ids.sub) <- levels(P10_krtno)
P10_krtno <- RenameIdents(P10_krtno, new.cluster.ids.sub)
```

##Fig Quality DimPlot
```{r fig.height=5, fig.width=7}
DimPlot(P10_krtno, label = TRUE, pt.size = 1.0, label.size = 6)#legend
DimPlot(P10_krtno, label = FALSE, pt.size = 1.0,
        cols = c('#f09527', '#ff692b', '#f0b627', '#0c3695', 
                 '#00206a', '#225ea8', '#1a9850', '#8c6bb1')) + NoLegend()#blank

```

#Keratinocyte Subset (Keratinocytes Only)
Choose supercluster and subset out contaminating cells
```{r}
cds_p10sub <- as.cell_data_set(P10_krtno, group.by = 'SCT_snn_res.0.15')#function from SeuratWrappers to convert to Monocle3 object, graph argument specifies the clustering to transfer (assay to use, and active resolution) to @clusters
#cds_p10sub <- cluster_cells(cds_p10sub, reduction_method = "UMAP", group_cells_by = 'seurat_clusters')#

#Add some info not carried over by SeuratWrappers
cds_p10sub <- estimate_size_factors(cds_p10sub)#add size factor metadata to CDS object
cds_p10sub@rowRanges@elementMetadata@listData[["gene_short_name"]] <- rownames(P10_krtno)#correct gene names in CDS object

#Validate UMAP coordinate transfer from Seurat object -> CDS object
temp1 <- P10_krtno@reductions$umap@cell.embeddings#large matrix with rows cell barcodes and 2 columns for the Seurat x/y UMAP coordinates
temp2 <- cds_p10sub@int_colData@listData$reducedDims$UMAP#matrix with rows cell barcodes and 2 columns for x/y UMAP coordinates
table(temp1 == temp2)#if all TRUE, seuratwrapper successfully transferred umap coordinates to CDS object via the graph argument
#cds_p10sub@int_colData@listData$reducedDims$UMAP <- P10_krtno@reductions$umap@cell.embeddings#not needed due to graph argument when making CDS object

#need to transfer Seurat PCA since this wasn't transferred by seurat wrapper
reducedDim(cds_p10sub, type = "PCA") <- P10_krtno@reductions$pca@cell.embeddings#transfer the matrix results of Seurat PCA to CDS
#cds_p10sub@preprocess_aux$prop_var_expl <- P10_krtno@reductions$pca@stdev#transfer numerical vector containing stdev values per PCA dimension

#partitions and cleaner subset still need to be created for learn_graph()
sub_cells <- choose_cells(cds_p10sub, return_list = TRUE)#get cell barcodes to make new seurat subset
plot <- DimPlot(P10_krtno, cells = as.vector(sub_cells))
plot
P10_krtno_sub <- subset(P10_krtno, cells = sub_cells)#subset to the clean clusters
```

Based on ElbowPlot, pick major PCs for next steps
#new UMAP
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
P10_krtno_sub <- RunUMAP(P10_krtno_sub, dims = 1:20, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
P10_krtno_sub <- FindNeighbors(P10_krtno_sub, dims = 1:20, verbose = FALSE)
P10_krtno_sub <- FindClusters(P10_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(P10_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P10_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
```{r fig.height=5, fig.width=5}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(P10_krtno_sub, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('KRT14', 'KRT5', 'KRT10', 'KRT18', 'KRT8', 'CALML5', 'CDSN', 'CNFN', 'CDH1',
         'TOP2A', 'ITGA6', 'ITGB1', 'SOX6', 'SOX9', 'COL17A1', 'DLL1', 'KRT15', 'EREG', 'nFeature_SCT', 'LEF1', 'EDAR')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P10_krtno_sub,
                    features = GOI[n_plot],
                    pt.size = 1.5,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
P10_krtno_sub <- FindClusters(P10_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.3)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P10_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P10_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

Save Seurat object
```{r}
save(P10_krtno_sub, file = '/t2tb/Object/P10_pig_krtno_sub.RData')
```

##Fig Quality DimPlot
```{r fig.height=5, fig.width=7}
path_out <- '~/Sean/UMAP/'
p1<- DimPlot(P10_krtno_sub, label = FALSE, pt.size = 1.5,
        cols = c('#f09527', '#ffab4b', '#ff692b', '#107de9', '#00206a', 
                 '#225ea8', '#f0b627', '#0c3695', '#d04800')) + NoLegend()#blank
p1

#ggsave(plot = p1, filename = 'Pig P10 UMAP (Krtno Sub).png', path = path_out,
#         width = 7, height = 7, units = 'in')
```

