---
title: "R Notebook"
output: html_notebook
---

#0. Load packages
```{r}
library(dplyr)
library(Seurat)
library(tictoc)
library(reticulate)
library(limma)#from Biocmanager
library(ggplot2)
library(viridisLite)
library(patchwork)
library(future)
plan('multisession', workers = 5)#cores = 5
options(future.globals.maxSize = 100000 * 1024^2)

```

#1. Load data
```{r}
# Load the D10 dataset
p3.data <- Read10X_h5("/media/sean/T2TB/Data/Thompson2025_PigP3/PigP3_scRNA/outs/p3pig.h5")

# Initialize the Seurat object with the raw (non-normalized data).
P3 <- CreateSeuratObject(counts = p3.data, project = "P3", min.cells = 3, min.features = 500)

P3
```

#QC
```{r}
# The [[ operator can add columns to object metadata. This is a great place to stash QC stats
P3[["percent.mt"]] <- PercentageFeatureSet(P3, pattern = "^MT")

# Visualize QC metrics as a violin plot
VlnPlot(P3, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)

# FeatureScatter is typically used to visualize feature-feature relationships, but can be used
# for anything calculated by the object, i.e. columns in object metadata, PC scores etc.

plot1 <- FeatureScatter(P3, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(P3, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```

#Determine QC cutoffs
```{r}
P3#before QC filter

VlnPlot(P3, features = 'nCount_RNA') + coord_cartesian(ylim = c(0, 20000))
VlnPlot(P3, features = 'nFeature_RNA') + coord_cartesian(ylim = c(0, 10000))

P3 <- subset(P3, subset = 
                 nFeature_RNA > 1000 &
                 nFeature_RNA < 10000 & 
                 nCount_RNA < 50000 &
                 percent.mt < 5)

P3
```

## SCTransform (Normalize, Scale Data)
```{r, warning=FALSE, message=FALSE}
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
P3 <- SCTransform(P3, vars.to.regress = c("percent.mt"), verbose = FALSE, do.scale = TRUE)
toc()
#Dimensional Reduction and Elbow Plot
P3 <- RunPCA(P3, verbose = FALSE)
ElbowPlot(P3, ndims = 50)
toc()
```

## UMAP (initial)
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
P3 <- RunUMAP(P3, dims = 1:30, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
P3 <- FindNeighbors(P3, dims = 1:30, verbose = FALSE)
P3 <- FindClusters(P3, verbose = FALSE, algorithm = 3)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P3, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P3, label = FALSE, pt.size = 1.0) + NoLegend()
#DimPlot(P3, label = FALSE)#no numbers on clusters, legend off to side

toc()
```

## FeaturePlots: view single-cell gene expression across clusters
```{r fig.height=5, fig.width=7, message=FALSE, warning=FALSE}
GOI <- c('nFeature_RNA','nCount_RNA', 'percent.mt',
         'PDGFRA', 'APCDD1', 'VIM', 'PHEROC', 'CHIA', 'ACTA2', 
         'CRABP1', 'ALX4', 'LEF1', 'RGS5', 'PDGFB', 'ANGPT1', 'CDH5', 'PECAM1', 'LYVE1', 'FLT4', 'PTPRC', 'CD86', 
         'CDH1', 'KRT14', 'KRT15', 'DLK2', 'NRG1', 'DLL1', 'TGFBI', 'TGFA', 'BMP7', 'LAMC1', 'COL17A1', 'PDGFC', 'PDGFA', 'KRT1', 'KRT10', 'SBSN', 'CALML5', 
         'SOX6', 'SOX9', 'KRT25', 'KRT73', 'KRT8', 'KRT18', 'TYROBP', 
         'ENSSSCG00000026302', 'TOP2A', 'ITGA6', 'FABP4', 'SOX10', 'AREG')
#path_out <- '.'
## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P3,
                    features = GOI[n_plot],
                    pt.size = 0.5,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
  }
toc()
```

### Change Resolution
```{r message=FALSE, warning=FALSE}
P3 <- FindClusters(P3, verbose = FALSE, algorithm = 3, resolution = 0.4)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P3, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P3, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

#Add rare celltypes as distinct clusters
TYROBP+/PTPRC- melanocytes vs TYROBP+/PTPRC+ immune cells; cluster 12 is 2 subclusters: dividing cells (KRT8+/KRT18+/TOP2A+) distinct from dividing blood vessels (PECAM1+/CDH5+/TOP2A+)
```{r fig.height=5, fig.width=15}
GOI <- c('TYROBP', 'PTPRC', 'KRT18', 'TOP2A', 'PECAM1', 'TOP2A')

#View Gene Co-Expression
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P3,
                    features = GOI[n_plot:(n_plot+1)],
                    pt.size = 0.1, order = TRUE, blend = TRUE, max.cutoff = 'q90')
  print(p1)
  n_plot <- n_plot + 2#iterate to the next 2-gene pairing
}

p1 <- DimPlot(P3, reduction = "umap")
ImmuneCells <- CellSelector(plot = p1)#fetch the immune cluster cell ids
DivMerkel <- CellSelector(plot = p1)#fetch the dividing merkel cell cluster ids

Idents(P3, cells = ImmuneCells) <- "17"
Idents(P3, cells = DivMerkel) <- "18"
```

#View the updated DimPlot
```{r fig.height=5, fig.width=7}
DimPlot(P3, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P3, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

## Differential Gene Expression
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Sean/DEG/P3/'#path to export folder
COI <- levels(P3)#define the clusters you wish to know diff. gene expression of

n_genes <- length(P3@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(P3, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, (n_loops-1), '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

## Label the Clusters by Cell Type
```{r message=FALSE, warning=FALSE}
levels(P3@active.ident)
table(P3@active.ident)
P3@active.ident <- factor(P3@active.ident, levels = c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
                                                      '11', '12', '13', '14', '15', '16', '17', '18'))
levels(P3@active.ident)
table(P3@active.ident)

P3$seurat_clusters <- P3@active.ident
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids <- c('0 Diff Krtno', '1 Basal Krtno', '2 Basal Krtno', '3 BV', '4 Sweat Gland',
                     '5 HF Krtno', '6 PF', '7 Pericyte', '8 Div Krtno', '9 SOX9 Krtno',
                     '10 DP', '11 HF Bulb', '12 Div BV', '13 Sweat Gland', '14 Melanocyte', 
                     '15 LV', '16 Nerve', '17 Immune Cell', '18 Div Sweat Gland')
names(new.cluster.ids) <- levels(P3)
P3 <- RenameIdents(P3, new.cluster.ids)
```

Save the Seurat object
```{r}
temp <- as.data.frame(strsplit2(P3@active.ident, split = ' '))
P3$seurat_clusters <- factor(temp$V1)
P3$seurat_clusters <- factor(P3$seurat_clusters, levels = c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9', '10',
                                                      '11', '12', '13', '14', '15', '16', '17', '18'))
save(P3, file = '/media/sean/T2TB/Object/p3_pig_allclusters_v2.RData')
```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Sean/UMAP/'
p1 <- DimPlot(P3, label = FALSE, pt.size = 1.0, 
              cols = c('#3b7bca', '#001952', '#002783', '#8c6bb1', '#08537e',
                     '#64a7e9', '#006837', '#7fcdbb', '#16539e', '#00206a',
                     '#00441b', '#96ccff', '#6f479c', '#1179a3', '#fc8d59', 
                     '#88419d', '#ffab4b', '#900101', '#0c6395')) + NoLegend()#manual colors w/o legend
p1
p1 <- DimPlot(P3, label = FALSE, pt.size = 1.0, 
              cols = c('#3b7bca', '#001952', '#002783', '#8c6bb1', '#08537e',
                     '#64a7e9', '#006837', '#7fcdbb', '#16539e', '#00206a',
                     '#00441b', '#96ccff', '#6f479c', '#1179a3', '#fc8d59', 
                     '#88419d', '#ffab4b', '#900101', '#0c6395'))#manual colors w/o legend
p1
#ggsave(plot = p1, filename = 'Pig P3 UMAP (All Clusters).png', path = path_out,
#         width = 7, height = 7, units = 'in')

```

```{r fig.height=7, fig.width=7}
#when adding immune and div sweat gland small clusters, the order of seurat cluster levels became 18, 17, 0:16, so the first relabeling was off by 2 positions
load('/media/sean/T2TB/Object/p3_pig_allclusters_v2.RData')

#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids <- c('18 Div Sweat Gland', '17 Immune Cell', '0 Diff Krtno', '1 Basal Krtno', '2 Basal Krtno', '3 BV', '4 Sweat Gland',
                     '5 HF Krtno', '6 PF', '7 Pericyte', '8 Div Krtno', '9 SOX9 Krtno',
                     '10 DP', '11 HF Bulb', '12 Div BV', '13 Sweat Gland', '14 Melanocyte', 
                     '15 LV', '16 Nerve')
names(new.cluster.ids) <- levels(P3)
P3 <- RenameIdents(P3, new.cluster.ids)

p1 <- DimPlot(P3, label = FALSE, pt.size = 1.0, 
              cols = c('#0c6395', '#900101', '#3b7bca', '#001952', '#002783', '#8c6bb1', '#08537e',
                     '#64a7e9', '#006837', '#7fcdbb', '#16539e', '#00206a',
                     '#00441b', '#96ccff', '#6f479c', '#1179a3', '#fc8d59', 
                     '#88419d', '#ffab4b')) + NoLegend()#manual colors w/o legend
p1
p1 <- DimPlot(P3, label = FALSE, pt.size = 1.0, 
              cols = c('#0c6395', '#900101', '#3b7bca', '#001952', '#002783', '#8c6bb1', '#08537e',
                     '#64a7e9', '#006837', '#7fcdbb', '#16539e', '#00206a',
                     '#00441b', '#96ccff', '#6f479c', '#1179a3', '#fc8d59', 
                     '#88419d', '#ffab4b'))#manual colors w/o legend
p1

save(P3, file = '/media/sean/T2TB/Object/p3_pig_allclusters_v3.RData')#fixed cluster name order
```


#2. Subset the Keratinocytes
```{r message=FALSE, warning=FALSE}
P3_krtno <- subset(P3, idents = c('0 Diff Krtno', '1 Basal Krtno', '2 Basal Krtno', '8 Div Krtno', '9 SOX9 Krtno'), invert = FALSE)#subset the keratinocyte supercluster
P3_krtno#view the subset

## We want to renormalize the data to bring out the heterogenetiy within the lineage
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
P3_krtno <- SCTransform(P3_krtno, vars.to.regress = "percent.mt", verbose = FALSE, do.scale = TRUE)
toc()
#Dimensional Reduction and Elbow Plot
P3_krtno <- RunPCA(P3_krtno, verbose = FALSE)
ElbowPlot(P3_krtno, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
P3_krtno <- RunUMAP(P3_krtno, dims = 1:20, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
P3_krtno <- FindNeighbors(P3_krtno, dims = 1:20, verbose = FALSE)
P3_krtno <- FindClusters(P3_krtno, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(P3_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P3_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
Includes some markers identified by Sole-Boldo et al. in their analysis as well as DAGs from Thompson et al. 2022
```{r fig.height=5, fig.width=7}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(P3_krtno, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('KRT14', 'KRT15', 'KRT5',  'KRT10', 'KRT1', 'SBSN', 'CALML5', 'CNFN',
         'SOX6', 'SOX9', 'KRT25', 'KRT73', 'KRT8', 'KRT18', 'TYROBP', 
         'ENSSSCG00000026302', 'TOP2A', 'ITGA6', 'ITGB4', 'CDH1', 'SOX10', 'AREG',
         'SOX6', 'SOX5', 'SOSTDC1', 'COL17A1',  'VEGFA', 'ANGPTL4', 'PDGFC', 'TGFBI', 'TGFB2', 'CXCL14', 
         'SOX9', 'TGFA', 'LEF1', 'DLK2', 'DLL1', 'JAG1', 'NOTCH3', 
         'DUSP6', 'CAV1', 'ASS1', 'VIM', 'PTPRC', 'PDGFRA', 'PECAM1', 'RGS5', 'percent.mt', 'nCount_SCT', 'AXIN2', 'WNT16')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P3_krtno,
                    features = GOI[n_plot],
                    pt.size = 1.0,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
P3_krtno <- FindClusters(P3_krtno, verbose = FALSE, algorithm = 3, resolution = 0.5)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P3_krtno, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P3_krtno, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

# Differential Gene Expression Within Keratinocytes
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Sean/DEG/Krtno_sub/P3_Krtno_Supercluster/'#path to export folder
COI <- c('0', '1', '2', '3', '4', '5', '6', '7', '8', '9')#define the clusters you wish to know diff. gene expression of

n_genes <- length(P3_krtno@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(P3_krtno, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, COI[n_loops], '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

## Label the clusters
```{r message=FALSE, warning=FALSE}
#updated renaming convention in Seurat 3.0.0+ and Signac 1.0.0+
new.cluster.ids.sub <- c('0', '1', '2', '3', '4', 
                         '5', '6', '7', '8', '9')
names(new.cluster.ids.sub) <- levels(P3_krtno)
P3_krtno <- RenameIdents(P3_krtno, new.cluster.ids.sub)
```

## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Desktop/Sean_CellChat/UMAP/'
p1 <- DimPlot(P3_krtno, label = FALSE, pt.size = 1.0, 
              cols = c('#f09527', '#f0b627', '#107de9', '#ff692b', 
                 '#2c72c8', '#0c3695', '#225ea8', '#8c6bb1', 
                 '#1a9850', '#00206a', '#d04800', '#6b92c3')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Pig P3 UMAP (Krtno).png', path = path_out,
         width = 7, height = 7, units = 'in')
```

#2b. Subset the Keratinocytes Only
```{r message=FALSE, warning=FALSE}
P3_krtno_sub <- subset(P3_krtno, idents = c('0', '1', '3', '4', 
                         '6', '8', '9'), invert = FALSE)#subset the IFE keratinocytes
P3_krtno_sub#view the subset

## We want to renormalize the data to bring out the heterogenetiy within the lineage
#Run sctransform (replaces NormalizeData, ScaleData, and FindVariableFeatures + RegressOut argument of ScaleData)
tic('Running SCTransform')
P3_krtno_sub <- SCTransform(P3_krtno_sub, vars.to.regress = "percent.mt", verbose = FALSE, do.scale = TRUE)
toc()
#Dimensional Reduction and Elbow Plot
P3_krtno_sub <- RunPCA(P3_krtno_sub, verbose = FALSE)
ElbowPlot(P3_krtno_sub, ndims = 50)
```

Based on ElbowPlot, pick major PCs for next steps
```{r message=FALSE, warning=FALSE}
tic('Running UMAPs')
#run the UMAP function using dimensions informed by elbow plot
P3_krtno_sub <- RunUMAP(P3_krtno_sub, dims = 1:20, verbose = FALSE, umap.method = "umap-learn", metric = "correlation")#default is dims = 1:30
#Higher PCs (dims=1:30) can represent subtle but relevant sources of heterogeneity
P3_krtno_sub <- FindNeighbors(P3_krtno_sub, dims = 1:20, verbose = FALSE)
P3_krtno_sub <- FindClusters(P3_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.6)#default is algorithm = 1 (Louvain), 3 = SLM
toc()
DimPlot(P3_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P3_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

### View Gene Expression of Keratinocyte Heterogeneity Markers
Includes some markers identified by Sole-Boldo et al. in their analysis as well as DAGs from Thompson et al. 2022
```{r fig.height=5, fig.width=7}
## FeaturePlots: view single-cell gene expression across clusters
DimPlot(P3_krtno_sub, label = TRUE, pt.size = 2, label.size = 6.0)

GOI <- c('KRT14', 'KRT15', 'KRT5',  'KRT10', 'KRT1', 'SBSN', 'CALML5', 'CNFN',
         'SOX6', 'SOX9', 'KRT25', 'ENSSSCG00000026302', 'TOP2A', 'ITGA6', 'ITGB4', 'CDH1', 'AREG',
         'SOSTDC1', 'COL17A1',  'VEGFA', 'ANGPTL4', 'PDGFC', 'TGFBI', 'TGFB2', 'CXCL14', 
         'SOX9', 'TGFA', 'LEF1', 'DLK2', 'DLL1', 'JAG1', 'NOTCH3', 'NRG1', 'OPCML', 'SOCS3',
         'DUSP6', 'CAV1', 'ASS1', 'VIM', 'percent.mt', 'nCount_SCT', 'AXIN2', 'WNT16')

## View Gene Expression ##
tic('FeaturePlot auto-export')
#Export loop for FeaturePlot() for each Gene in GOI list
n_plot <- 1
while (n_plot < length(GOI) + 1) {
  p1 <- FeaturePlot(object = P3_krtno_sub,
                    features = GOI[n_plot],
                    pt.size = 1.0,
                    order = TRUE) + 
    scale_colour_gradientn(colours = magma(50))#iterate through the genes of interest vector inputting their coordinates
  
  print(p1)#spare the markdown file saving tons of plots
  
  #save the FeaturePlots automatically
  #ggsave(plot = p1, filename = paste(GOI[n_plot], 'FeaturePlot.png', sep = ' '), path = path_out,
  #       width = 6, height = 4, units = 'in')
  n_plot <- n_plot + 1
}
toc()

```

### Change Resolution
```{r message=FALSE, warning=FALSE}
P3_krtno_sub <- FindClusters(P3_krtno_sub, verbose = FALSE, algorithm = 3, resolution = 0.5)#default is algorithm = 1 (Louvain), 3 = SLM
DimPlot(P3_krtno_sub, label = TRUE, pt.size = 1.0, label.size = 6) + NoLegend()#numbers on clusters
DimPlot(P3_krtno_sub, label = FALSE, pt.size = 1.0) + NoLegend()#no labels (to label in photoshop)
```

# Differential Gene Expression Within Keratinocytes
```{r}
tic('ClusterMarker auto-export loop')#start timer
path_out <- '~/Sean/DEG/Krtno_sub/P3_Krtno_sub1/'#path to export folder
COI <- c('0', '1', '2', '3', '4', '5', '6', '7', '8')#define the clusters you wish to know diff. gene expression of

n_genes <- length(P3_krtno_sub@assays[["SCT"]]@data@Dimnames[[1]])#get # of genes we will be testing
n_loops <- 1
while (n_loops < length(COI) + 1) {
#Find positive markers (upregulated genes)
cluster.markers.temp <- FindMarkers(P3_krtno_sub, ident.1 = COI[n_loops], min.pct = 0.05, logfc.threshold = 0.0, assay = 'SCT', only.pos = TRUE)#iterate through the entire COI list
#head(cluster.markers.temp, n = 30)
cluster.markers.temp$Gene = rownames(cluster.markers.temp)#add a column for the gene name
cluster.markers.temp$pct.2[cluster.markers.temp$pct.2 == 0] <- NA#corrects pct ratio error if pct.2=0
cluster.markers.temp <- na.omit(cluster.markers.temp)#remove rows with NA
cluster.markers.temp$Pct_Ratio = cluster.markers.temp$pct.1 / cluster.markers.temp$pct.2#compute ratio of how much the cluster of interest expresses compared to all other cells
cluster.markers.temp <- cluster.markers.temp %>% dplyr::select('Gene', everything())#move gene column to the front of the dataframe
cluster.markers.temp$'PctRatio_x_logfc' = cluster.markers.temp$Pct_Ratio * cluster.markers.temp$avg_log2FC
cluster.markers.temp <- cluster.markers.temp[cluster.markers.temp$PctRatio_x_logfc > quantile(cluster.markers.temp$PctRatio_x_logfc, 0.95), ]#keep only the top 95% of upregulated genes in ident.1
cluster.markers.temp <- cluster.markers.temp[order(cluster.markers.temp$PctRatio_x_logfc, decreasing = TRUE), ]#rank genes
write.csv(cluster.markers.temp, file = paste(path_out, COI[n_loops], '_pos_markers.csv', sep = ''), row.names = FALSE)#save that cluster's markers

n_loops <- n_loops + 1#track for when all clusters of one dataset have been run through
}
toc()#end timer
```

Save the Seurat object
```{r}
save(P3_krtno_sub, file = '/media/sean/T2TB/Object/P3_pig_krtno_sub1.RData')
```


## Recolor the UMAP clusters by cell lineage
```{r fig.height=7, fig.width=7}
path_out <- '~/Sean/UMAP/'
p1 <- DimPlot(P3_krtno_sub, label = FALSE, pt.size = 2.0, 
              cols = c('#f0b627', '#002783', '#225ea8', '#0c3695', '#08537e', 
                       '#001952', '#00206a', '#1179a3', '#ff692b')) + NoLegend()#manual colors w/o legend
p1
ggsave(plot = p1, filename = 'Pig P3 UMAP (Krtno sub1).png', path = path_out,
         width = 7, height = 7, units = 'in')

```
